cli/global/setup/setup.go

```diff
@@ -0,0 +1,126 @@
+package global_setup
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/kubeshop/tracetest/cli/analytics"
+	"github.com/kubeshop/tracetest/cli/config"
+	global_parameters "github.com/kubeshop/tracetest/cli/global/parameters"
+	misc_actions "github.com/kubeshop/tracetest/cli/misc/actions"
+	"github.com/kubeshop/tracetest/cli/utils"
+	"github.com/spf13/cobra"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+type Setup struct {
+	Config      *config.Config
+	Logger      *zap.Logger
+	VersionText *string
+	Parameters  *global_parameters.Global
+	options     []SetupOption
+}
+
+type SetupOption = func(setup *Setup, args []string) error
+
+func NewSetup(parameters *global_parameters.Global, options ...SetupOption) *Setup {
+	return &Setup{
+		Parameters: parameters,
+		options:    options,
+	}
+}
+
+func (s *Setup) PreRun(cmd *cobra.Command, args []string) (string, error) {
+	for _, option := range s.options {
+		err := option(s, args)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return "", nil
+}
+
+func WithInitAnalytics() SetupOption {
```

Comment:
Thinking about the `Teardown` discussion: this option feels more to a decorator that uses a resource for the entire command run than just an option. Maybe we could decorate commands that uses analytics and logger to have the startup/teardown logic encapsulated in one place. It makes sense?

Reply:
I was thinking on something like this:

```go
func DecorateCommandWithAnalytics(cmd *cobra.Command) cmd {
    // decorate PreRun
    // decorate PostRun

    return decoratedCommand
}
```

Reply:
Makes sense, I'm going to work on an update on this between hackathon off time haha